File: nodejs/land.js
const { FileSystemWallet, Gateway } = require('fabric-network');
const path = require('path');

const ccpPath = path.resolve(__dirname, '..', 'connection', 'connection-org1.json');

/**
 * 提供操作区块链链接
 * 
 * @param {string} userId 钱包中的用户，用于验证身份
 * @param {*} infomation 操作的信息，如参数、用户、执行的动作
 * @param {*} isQuery 是否是一个查询动作
 * @param {*} contractFile 合同 [默认土地合同(在本文件中)]
 * @returns {Array} 返回一个二位数组，第一个元素为结果，第二个元素是错误信息
 */
async function baseConnection(userId, infomation, isQuery = true, contractFile = 'land') {
    try {

        // Create a new file system based wallet for managing identities.
        const walletPath = path.join(process.cwd(), 'wallet');
        const wallet = new FileSystemWallet(walletPath);
        console.log(`Wallet path: ${walletPath}`);

        // Check to see if we've already enrolled the user.
        const userExists = await wallet.exists(userId);
        if (!userExists) {
            console.log('An identity for the user "user1" does not exist in the wallet');
            console.log('Run the registerUser.js application before retrying');
            throw new Error("user does not register.");
            // return;
        }

        // Create a new gateway for connecting to our peer node.
        const gateway = new Gateway();
        await gateway.connect(ccpPath, { wallet, identity: userId, discovery: { enabled: true, asLocalhost: true } });

        // Get the network (channel) our contract is deployed to.
        const network = await gateway.getNetwork('mychannel');

        // Get the contract from the network.
        const contract = network.getContract(contractFile);

        let result = null;

        if (isQuery)
            result = await contract.evaluateTransaction(infomation.method, ...infomation.args);
        else
            result = await contract.submitTransaction(infomation.method, ...infomation.args);

        // Disconnect from the gateway.
        gateway.disconnect();

        return [result?.toString(), null];
    } catch (error) {
        console.error(`Failed to submit transaction: ${error}`);
        return [null, error?.toString()];
    }

}




/*********************< LAND >*********************************/

// invoke
/**
 * 使用参数创建土地
 */
async function CreateLand(infomation) {
    return await baseConnection(infomation.userId, infomation, false);
}

/**
 * 管理员通过 土地上传申请
 * @param args [landId]
 */
async function ValidLand(infomation) {
    return await baseConnection(infomation.userId, infomation, false);
}

/**
 * 使用键值更改土地信息
 * @param args [landId, key, value]
 * 例如: ['beijing0101', 'owner', 'pinxue02']
 */
async function UpdateLand(infomation) {
    return await baseConnection(infomation.userId, infomation, false);
}


// query
/**
 * 使用 landId 查询土地信息，返回土地信息对象
 * @param args [landId]
*/
async function QueryLand(infomation) {
    return await baseConnection(infomation.userId, infomation);
}

/**
 * 使用 其他键值 查询土地信息/列表
 * @param arg [key, value]
 * 例如：
 *    1. [CreatedBy, pinxue01]
 *    2. [Position, beijing]
 */
async function QueryLandByKey(infomation) {
    return await baseConnection(infomation.userId, infomation);
}

module.exports.Land = {
    CreateLand,
    ValidLand,
    UpdateLand,
    QueryLand,
    QueryLandByKey
}




/****************************< USRE >********************************/

// invoke
/**
 * @param args [name, userId, password, role, whoboss] 
 */
async function CreateUser(infomation) {
    return await baseConnection(infomation.userId, infomation, false, 'users');
}

// query 
/**
 * [userId, password]
 */
async function LogIn(infomation) {
    return baseConnection(infomation.userId, infomation, true, 'users');
}

/**
 * [userId]
 */
async function QueryUser(infomation) {
    return baseConnection(infomation.userId, infomation, true, 'users');
}

module.exports.User = {
    CreateUser,
    LogIn,
    QueryUser
}



/*****************************< TRANSATION >****************************/

//  invoke
/**
 * @param arg [userId, transationId, landId, requester, validar, isValid, date, price]
 */
async function CreateTransation(infomation) {
    return baseConnection(infomation.userId, infomation, false, 'tran');
}

/**
 * @param arg [transationId]
 */
async function ValidTransation(infomation) {
    return baseConnection(infomation.userId, infomation, false, 'tran');
}

// query
/**
 * @param arg [transationId]
 */
async function QueryTransation(infomation) {
    return baseConnection(infomation.userId, infomation, true, 'tran');
}

/**
 * @param arg [key, value]
 */
async function QueryTransationByKey(infomation) {
    return baseConnection(infomation.userId, infomation, true, 'tran');
}


module.exports.Transation = {
    CreateTransation,
    ValidTransation,
    QueryTransation,
    QueryTransationByKey
}

/**********************< END >**************************/

File: main/land.go
package main

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

/***********************  invoke  ******************************/

func CreateLand(c *gin.Context) {
	network := GetNetwork()

	var body struct {
		Position      string `json:"position" binding:"required"`
		LandId        string `json:"landId" binding:"required"`
		Owner         string `json:"owner" binding:"required"`
		Valid         string `json:"valid" binding:"required"`
		InTransaction string `json:"inTransaction" binding:"required"`
	}

	// 检查字段
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": "fail",
			"data":   nil,
			"info":   err.Error(),
		})
		return
	}

	// 检查ID是否冲突
	// fixme: 应该放在链码内部
	sameLand, _ := BaseQuery(network, Information{"land", "QueryLand", []string{body.LandId}})
	if sameLand != "" {
		c.JSON(http.StatusBadRequest, Response{"fail", "The land already exist.", nil})
		return
	}

	// 上链
	result, err := BaseInvoke(network, Information{"land", "CreateLand", []string{
		body.Position, body.LandId, body.Owner, body.Valid, body.InTransaction,
	}})
	if err != nil {
		c.JSON(http.StatusBadRequest, Response{"fail", err.Error(), nil})
		return
	}

	c.JSON(http.StatusOK, Response{"ok", "", result})
}

func ValidLand(c *gin.Context) {
	network := GetNetwork()

	landId, isMatch := c.Params.Get("landId")
	if isMatch {
		result, err := BaseInvoke(network, Information{"land", "ValidLand", []string{landId}})
		if err != nil {
			c.JSON(http.StatusBadRequest, Response{"fail", err.Error(), nil})
			return
		}

		c.JSON(http.StatusOK, Response{"ok", "", result})
	} else {
		c.JSON(http.StatusBadRequest, Response{"fail", "land ID is required.", nil})
	}
}

func UpdateLand(c *gin.Context) {
	network := GetNetwork()

	var body struct {
		LandId string `json:"landId" binding:"required"`
		Key    string `json:"key" binding:"required"`
		Value  string `json:"value" binding:"required"`
	}

	// 检查字段
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status": "fail",
			"data":   nil,
			"info":   err.Error(),
		})
		return
	}

	result, err := BaseInvoke(network, Information{"land", "UpdateLand", []string{body.LandId, body.Key, body.Value}})
	if err != nil {
		c.JSON(http.StatusBadRequest, Response{"fail", err.Error(), nil})
		return
	}

	c.JSON(http.StatusOK, Response{"ok", "", result})
}

/***********************   query  ******************************/

func QueryLand(c *gin.Context) {
	network := GetNetwork()

	landId, isMatch := c.Params.Get("landId")
	if isMatch {
		result, err := BaseQuery(network, Information{"land", "QueryLand", []string{landId}})

		if err != nil {
			c.JSON(http.StatusBadRequest, Response{"fail", err.Error(), nil})
			return
		}
		c.JSON(http.StatusOK, Response{"ok", "", result})
	} else {
		c.JSON(http.StatusBadRequest, Response{"fail", "land id is required.", nil})
	}
}

func QueryLandByKey(c *gin.Context) {
	network := GetNetwork()

	key := c.Query("key")
	value := c.Query("value")

	if key != "" && value != "" {
		// 查链
		result, err := BaseQuery(network, Information{"land", "QueryLandByKey", []string{key, value}})

		if err != nil {
			c.JSON(http.StatusBadRequest, Response{"fail", err.Error(), nil})
			return
		}

		c.JSON(http.StatusOK, Response{"ok", "", result})
	} else {
		c.JSON(http.StatusBadRequest, Response{"fail", "key and value is not a pair.", nil})
	}
}
